k8s_events_mvp.md

```diff
@@ -0,0 +1,54 @@
+
+### Implementation Details
+
+Here's a sample event on a Tilt deployment of servantes' fe service:
+```
+(*v1.Event)(0xc000cdf400)(&Event{ObjectMeta:k8s_io_apimachinery_pkg_apis_meta_v1.ObjectMeta{Name:matt-fe-649447bf8.159eedef32f08d04,GenerateName:,Namespace:default,SelfLink:/api/v1/namespaces/default/events/matt-fe-649447bf8.159eedef32f08d04,UID:6da2ea3d-773d-11e9-8112-025000000001,ResourceVersion:929420,Generation:0,CreationTimestamp:2019-05-15 14:15:32 -0400 EDT,DeletionTimestamp:<nil>,DeletionGracePeriodSeconds:nil,Labels:map[string]string{},Annotations:map[string]string{},OwnerReferences:[],Finalizers:[],ClusterName:,Initializers:nil,ManagedFields:[],},InvolvedObject:ObjectReference{Kind:ReplicaSet,Namespace:default,Name:matt-fe-649447bf8,UID:c9bcc82c-773c-11e9-8112-025000000001,APIVersion:extensions,ResourceVersion:929414,FieldPath:,},Reason:SuccessfulDelete,Message:Deleted pod: matt-fe-649447bf8-77hgq,Source:EventSource{Component:replicaset-controller,Host:,},FirstTimestamp:2019-05-15 14:15:32 -0400 EDT,LastTimestamp:2019-05-15 14:15:32 -0400 EDT,Count:1,Type:Normal,EventTime:0001-01-01 00:00:00 +0000 UTC,Series:nil,Action:,Related:nil,ReportingController:,ReportingInstance:,})
+```
+
+#### Mapping an event to a manifest name
+
+The chief complexity here is that when we get an event, we want to know which manifest's log it should appear in.
+We inject into all deployed objects a `tilt-manifest` label that tells us its manifest. This does not show up in the event. Thus, we need to get from the event to its associated k8s object, so we can check its labels. There are two fields we could use for this: `UID` and `ObjectReference`. Tilt does not currently store the appropriate information to determine the manifest from either of these values.
+
+We've discussed, for other purposes, changing our `kubectl apply` to use `-ojson` to get UIDs back for the objects we create. Unfortunately, that doesn't suffice for this case. The event is on the ReplicaSet created by the Deployment. Running `kubectl apply -ojson` gives us the Deployment's UID, but not the ReplicaSet's.
+
+I think we want something like a `map[model.UID]model.ManifestName` or `map[v1.ObjectReference]model.ManifestName` that we could use to determine to which manifest we should log an Event.
+
+##### Option 1 - watch all events and keep the map up-to-date
+Ideally, we could just watch all objects and update the map as they're added/deleted, but that doesn't seem to be possible via the k8s api (e.g., [1](https://github.com/kubernetes/kubernetes/pull/57792), [2](https://github.com/kubernetes/kubernetes/issues/1685)).
+
+We could set up individual watches for all object types to get the same effect. Maybe we'll want to do that eventually, but it feels dirty. This means: 1) getting a list of the potential object types (and, really, we should also set up a watcher on the defined object types to ensure we pick up any new CRDs), 2) looping over those and creating a watcher on each.
```

Comment:
I think it's good to watch lots of things instead of just everything so that we can limit our watching (e.g. do we want each Tilt user to be seeing every event on some huge cluster?)

Having it be a (nearly) constant number of watches doesn't concern me greatly, as compared to one watch on an unbounded number of objects.

Reply:
Sorry, I used sloppy language here. This is proposing, e.g.:
```
manifestsByUID := make(map[UID]ManifestName)
k8sTypes := []string{'pods', 'deployments', 'statefulsets', 'replicasets', ...}
for _, k8sType in range k8sTypes {
  go func() {
    for e := <- range k8sClient.Watch(k8sType) {
      manifestsByUID[e.UID] = e.labels[TiltManifestLabelName]
    }
  }
} 
```

i.e., we call the k8s api once per k8s resource type to be notified of creation/deletion of all objects of that type

(unfortunately, the k8s api has at least two types named Event: [as used in the name of this feature](https://github.com/windmilleng/tilt/blob/38b55952ab972efd330a1adaac710540531db288/vendor/k8s.io/api/core/v1/types.go#L4631) and [individual notifications in the watch API](https://github.com/windmilleng/tilt/blob/88718afdd5f4d745ce461cb7cc9bbcd6d825f316/vendor/k8s.io/apimachinery/pkg/watch/watch.go#L54))

I'm not sure I understand what distinction you're trying to draw or what APIs you're talking about using, or if this is just all confusion caused by my sloppy language, but AFAICT, if we want ensure we can map events to manifests, we'd need to call the watch api once per k8s resource type, and receive notifications for *all* objects in the cluster.

We could slim that down by filtering for the Tilt Run ID label, but 1) then we're making things hard for CRDs again (though maybe we just drop in extra_pod_labels, which seems reasonable) and 2) we've been talking about moving away from that, but maybe it's a mistake to hedge on a thing that's just talk.